xdsl/transforms/reconcile_unrealized_casts.py

```diff
@@ -0,0 +1,117 @@
+import itertools
+from collections import deque
+from typing import Iterable
+from warnings import warn
+
+from xdsl.dialects import builtin
+from xdsl.dialects.builtin import ModuleOp
+from xdsl.ir import MLContext
+from xdsl.ir.core import Use
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+
+class ReconcileUnrealizedCasts(RewritePattern):
+    """
+    An `unrealized_conversion_cast` operation represents an unrealized conversion from one set of types to another,
+    that is used to enable the inter-mixing of different type systems.
+
+    This pattern rewriter removes `unrealized_conversion_cast` operations that are not needed.
+    """
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: builtin.UnrealizedConversionCastOp, rewriter: PatternRewriter
+    ) -> None:
+        # We remove several casts in a single pass, which should not happen
```

Comment:
Are we sure we want this behaviour? The similar pass in the workshop-all branch only modifies one cast at a time, and the only difference seems to be that it changes cast chains like `abcd` to `ad`, unlike MLIR. I would argue that if we want to go with the MLIR approach then we should write this not as a rewrite pattern but a straight-up function that iterates over the IR itself.

Reply:
I've just done it, and if that's alright with you I think we can merge @superlopuh ?